### PR TITLE
refactor: Update PHPDoc Common::config

### PR DESCRIPTION
**Pull request by maniaba**
**Description**
The `@template` annotation relates to a concept called Generics, which does not currently exist in PHP, but are a way of dynamically describing the contents of a parameter or return type that would be unknown until a class is instantiated or method called.

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Common.php [hunk 1]
```diff
@@ -973,10 +978,9 @@ function route_to(string $method, ...$params)
      *    session()->set('foo', 'bar');
      *    $foo = session('bar');
      *
-     * @param string $val
+     * @param string|null $val
      *
      * @return array|bool|float|int|object|Session|string|null
-     * @phpstan-return ($val is null ? Session : array|bool|float|int|object|string|null)
```
Comment (kenjis) [inline system/Common.php:979]: Why do you remove this?

**Comment by kenjis** [comment]
Thank you for sending PR.

Please sign all your commits:
https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#signing

Fix the coding style:
https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#php-style

Fix PHPStan errors:
https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#static-analysis-on-php-code

File: system/Common.php [hunk 2]
```diff
@@ -203,7 +204,12 @@ function command(string $command)
     /**
      * More simple way of getting config instances from Factories
      *
-     * @return object|null
+     * @template ConfigTemplate of BaseConfig
+     *
+     * @param class-string<ConfigTemplate>|string $name
+     *
+     * @return ConfigTemplate
+     * @phpstan-return ($name is class-string<ConfigTemplate> ? ConfigTemplate : object|null)
```
Comment (paulbalandan) [inline system/Common.php:212]: Does it really return a generic object? I thought if not found the function will return `null`.

Comment (maniaba) [reply system/Common.php:212]: I add null as a possible return

File: system/Common.php [hunk 3]
```diff
@@ -203,7 +204,12 @@ function command(string $command)
     /**
      * More simple way of getting config instances from Factories
      *
-     * @return object|null
+     * @template ConfigTemplate of BaseConfig
+     *
+     * @param class-string<ConfigTemplate>|string $name
```
Comment (kenjis) [inline system/Common.php:209]: Is this better according to our current rule?
Probably the type `ConfigTemplate` is not understood by IDEs.

```suggestion
     * @phpstan-param class-string<ConfigTemplate>|string $name
```

File: system/Common.php [hunk 4]
```diff
@@ -203,7 +204,12 @@ function command(string $command)
...
+     *
+     * @return ConfigTemplate|null
```
Comment (kenjis) [inline system/Common.php:211]: Like this?
```suggestion
     * @return BaseConfig|null
```

Comment (kenjis) [inline system/Common.php:211]: Probably this must be reverted.

```suggestion
     * @return object|null
```

File: system/Common.php (hunk 3 above)
Comment (maniaba) [reply system/Common.php:209]: @kenjis After this change phpstan throws the errors.

Comment (kenjis) [reply system/Common.php:209]: What errors?
No errors on my local environment.
```diff
--- a/system/Common.php
+++ b/system/Common.php
@@ -206,9 +206,9 @@ if (! function_exists('config')) {
      *
      * @template ConfigTemplate of BaseConfig
      *
-     * @param class-string<ConfigTemplate>|string $name
+     * @phpstan-param class-string<ConfigTemplate>|string $name
      *
-     * @return ConfigTemplate|null
+     * @return object|null
      * @phpstan-return ($name is class-string<ConfigTemplate> ? ConfigTemplate : object|null)
      */
     function config(string $name, bool $getShared = true)
```

Comment (maniaba) [reply system/Common.php:209]: When we use @return object|null and only @phpstan-param there is no auto suggestion (is not understood by IDEs).

Comment (kenjis) [reply system/Common.php:209]: @maniaba Exactly, PhpStorm does not do auto suggestion.

**Review by kenjis — APPROVED** [review]
This is the same thing as `model()`, and works with PhpStorm.

File: system/Common.php (hunk 3 above)
Comment (maniaba) [reply system/Common.php:209]: let's leave it as it is right now?
It is good for phpstan and IDE.

**Review by paulbalandan — APPROVED** [review]

**Comment by kenjis** [comment]
@maniaba Thank you.

We will accept the change from `config('Foo')` to `config(Foo::class)` in everywhere.

**Pull request merged**